src/web/releases.rs

```diff
@@ -244,57 +244,94 @@ fn get_releases_by_owner(
     (author_name, packages)
 }
 
+/// Get the search results for a search query
+///
+/// Retrieves crates which names have a levenshtein distance of less than or equal to 3,
+/// crates who fit into or otherwise are made up of the query or crates who's descriptions
+/// match the search query.
+///
+/// * `query`: The query string, unfiltered
+/// * `page`: The page of results to show (1-indexed)
+/// * `limit`: The number of results to return
+///
+/// Returns `None` if no results are found and `Some` with the total number of results and the
+/// currently requested results
+///
 fn get_search_results(
     conn: &Connection,
-    query: &str,
+    mut query: &str,
     page: i64,
     limit: i64,
 ) -> Option<(i64, Vec<Release>)> {
+    query = query.trim();
+    let split_query = query.replace(' ', " & ");
     let offset = (page - 1) * limit;
-    let mut packages = Vec::new();
 
-    let rows = match conn.query(
-        "SELECT crates.name,
-                                    releases.version,
-                                    releases.description,
-                                    releases.target_name,
-                                    releases.release_time,
-                                    releases.rustdoc_status,
-                                    ts_rank_cd(crates.content, to_tsquery($1)) AS rank
-                                 FROM crates
-                                 INNER JOIN releases ON crates.latest_version_id = releases.id
-                                 WHERE crates.name LIKE concat('%', $1, '%')
-                                    OR crates.content @@ to_tsquery($1)
-                                 ORDER BY crates.name = $1 DESC,
-                                    crates.name LIKE concat('%', $1, '%') DESC,
-                                    rank DESC
-                                 LIMIT $2 OFFSET $3",
-        &[&query, &limit, &offset],
-    ) {
-        Ok(r) => r,
-        Err(_) => return None,
-    };
+    let rows = conn
+        .query(
+            "SELECT crates.name,
+                MAX(releases.version) AS version,
+                MAX(releases.description) AS description,
```

Comment:
You can add a `WHERE` clause like

```sql
releases.id = (
  SELECT releases.id
  FROM releases
  WHERE
    releases.crate_id = crates.id
    AND releases.rustdoc_status
    AND NOT releases.yanked
  ORDER BY releases.version DESC
  LIMIT 1
)
```

and then add `releases.id` to the `GROUP BY` since there'll only be a distinct row. (That also takes care of filtering out crates that don't have successful builds since it'll return NULL and not match any rows).

Reply:
Oh, you'll also have to change `SUM(downloads)` to `downloads_total` since there won't be all the releases to sum over anymore.

Reply:
Hey, you can use the array thingy to get the latest value like this:

```sql
MAX(ARRAY([EXTRACT('EPOCH' FROM releases.release_time)::text, releases.target_name]))[1] AS target_name,
```

But that's really a nasty hack. Looks like what you really need is another table with the latest release for each crate:

```sql

SELECT 
    crates.name,
    latest_release.version AS version,
    latest_release.description AS description,
    latest_release.target_name AS target_name,
    latest_release.release_time AS release_time,
    latest_release.rustdoc_status AS rustdoc_status
    crates.github_stars,
    SUM(releases.downloads) AS downloads,
    -- The levenshtein distance between the search query and the crate's name
    levenshtein_less_equal($1, crates.name, 3) as distance,
    -- The similarity of the tokens of the search vs the tokens of `crates.content`.
    -- The `32` normalizes the number by using `rank / (rank + 1)`
    ts_rank_cd(crates.content, to_tsquery($2), 32)  as content_rank
FROM 
    crates 
    INNER JOIN releases.crate_id = crates.id
    INNER JOIN (
        SELECT DISTINCT ON (crate_id)
            crate_id,
            version,
            description,
            target_name,
            release_time,
            rustdoc_status
        FROM 
            releases
        ORDER BY
            create_id,
            release_time DESC
    ) AS latest_release ON latest_release.crate_id = crates.id
```
I haven't tested any of this and i'm not familiar with the tables. Also, it looks like you are missing a group by.

Reply:
I think that's a great idea! Shouldn't the inner query have `LIMIT 1` though since we only want the latest?

Reply:
Notice the DISTINCT ON in the latest_release subquery. It will return only the first row for each crate. The order is determined by the order by, so it will return the latest. 

You can an example here:
https://hakibenita.com/the-many-faces-of-distinct-in-postgre-sql#distinct-on

Reply:
Well, all modifications included, I've got this query
```sql
SELECT
    crates.name,
    latest_release.version AS version,
    latest_release.description AS description,
    latest_release.target_name AS target_name,
    latest_release.release_time AS release_time,
    latest_release.rustdoc_status AS rustdoc_status,
    crates.github_stars,
    SUM(releases.downloads) AS downloads,
    -- Get the total number of results, disregarding the limit
    COUNT(*) OVER() as total,

    -- The levenshtein distance between the search query and the crate's name
    levenshtein_less_equal(CAST($1 AS TEXT), CAST(crates.name AS TEXT), 3) as distance,
    -- The similarity of the tokens of the search vs the tokens of `crates.content`.
    -- The `32` normalizes the number by using `rank / (rank + 1)`
    ts_rank_cd(crates.content, plainto_tsquery($1), 32)  as content_rank
FROM
    crates
    INNER JOIN releases.crate_id = crates.id
    INNER JOIN (
        SELECT DISTINCT ON (crate_id)
            crate_id,
            version,
            description,
            target_name,
            release_time,
            rustdoc_status
        FROM
            releases
        ORDER BY
            create_id,
            release_time DESC
    ) AS latest_release ON latest_release.crate_id = crates.id

-- Filter crates that haven't been built and crates that have been yanked and
-- crates that don't match the query closely enough
WHERE
    latest_release.rustdoc_status
    AND NOT latest_release.yanked
    AND (
        -- Crates names that match the query sandwiched between wildcards will pass
        crates.name ILIKE CONCAT('%', $1, '%')
        -- Crate names with which the levenshtein distance is closer or equal to 3 will pass
        OR levenshtein_less_equal($1, crates.name, 3) <= 3
        -- Crates where their content matches the query will pass
        OR plainto_tsquery($1) @@ crates.content
    )

GROUP BY crates.id, releases.id

-- Ordering is prioritized by how closely the query matches the name, how closely the
-- query matches the description finally how many downloads the crate has
ORDER BY
    distance ASC,
    content_rank DESC,
    downloads_total DESC

-- Allows pagination
LIMIT $2 OFFSET $3
```
but it seems that the `INNER JOIN releases.crate_id = crates.id` is a syntax error?

Reply:
```sql
INNER JOIN releases ON releases.crate_id = crates.id
```

Reply:
You are also missing `latest_release.yanked`. add `yanked` to the `latest_release` subquery.